query.go

```diff
@@ -111,145 +121,128 @@ loop:
-		// which get queried from among the unqueried K  closet.
-		peersToQuery := q.localPeers.UnqueriedFromKClosest(q.scorePeerByDistanceAndLatency,
-			func(i1 peerheap.Item, i2 peerheap.Item) bool {
-				return i1.Value.(*big.Int).Cmp(i2.Value.(*big.Int)) == -1
-			})
-
-		// The lookup terminates when the initiator has queried and gotten responses from the k
-		// closest nodes it has heard about.
-		if len(peersToQuery) == 0 {
+		select {
+		case sawPeers := <-ch:
+			q.populatePeers(<-ch)
+		case <-pathCtx.Done():
+			q.terminate()
 			return
 		}
 
-		// Of the k nodes the initiator has heard of closest to the target,
-		// it picks α that it has not yet queried and resends the FIND NODE RPC to them.
-		numQuery := q.dht.alpha
-
-		// However, If a round of RPC's fails to return a node any closer than the closest already heard about,
-		// the initiator resends the RPC'S to all of the k closest nodes it has
-		// not already queried.
-		if !foundCloser {
-			numQuery = len(peersToQuery)
-		} else if pqLen := len(peersToQuery); pqLen < numQuery {
-			// if we don't have α peers, pick whatever number we have.
-			numQuery = pqLen
+		// termination is triggered on end-of-lookup conditions or starvation of unused peers
+		if q.readyToTerminate() {
+			q.terminate()
+			return
 		}
 
-		// reset foundCloser to false for the next round of RPC's
-		foundCloser = false
-
-		queryResCh := make(chan *queryResult, numQuery)
-		resultsReceived := 0
+		// if all "threads" are busy, wait until someone finishes
+		if q.queryPeers.NumWaiting() >= alpha {
+			continue
+		}
 
-		// send RPC's to all the chosen peers concurrently
-		for _, p := range peersToQuery[:numQuery] {
-			go func(p peer.ID) {
-				queryResCh <- q.queryPeer(p)
-			}(p)
+		// spawn new queries, up to the parallelism allowance
+		for j := 0; j < alpha-q.queryPeers.NumWaiting(); j++ {
+			q.spawnQuery(ch)
 		}
+	}
+}
 
-	loop:
-		// wait for all outstanding RPC's to complete before we start the next round.
-		for {
-			select {
-			case res := <-queryResCh:
-				foundCloser = foundCloser || res.foundCloserPeer
-				resultsReceived++
-				if resultsReceived == numQuery {
-					break loop
-				}
-			case <-q.ctx.Done():
-				return
-			}
+// spawnQuery starts one query, if an available seen peer is found
+func (q *query) spawnQuery(ch chan<- *queryUpdate) {
+	if peers := q.queryPeers.GetSortedSeen(); len(peers) == 0 {
+		return
+	} else {
+		q.queryPeers.MarkWaiting(peers[0])
+		go q.queryPeer(ch, peers[0])
+	}
+}
+
+func (q *query) readyToTerminate() bool {
+	// if termination has already been determined, the query is considered terminated forever,
+	// regardless of any change to queryPeers that might occur after the initial termination.
+	if q.terminated {
+		return true
+	}
+	// give the application logic a chance to terminate
+	if q.stopFn(q.queryPeers) {
+		return true
+	}
+	if q.isStarvationTermination() {
+		return true
+	}
+	if q.isLookupTermination() {
+		return true
+	}
```

Comment:
Doesn't StarvationTermination imply LookupTermination? If no peer is in the Seen or Waiting state it means they're all Queried or Unreachable which means that there is a LookupTermination.

Reply:
What _could_ happen is that you could have talked to reachable peers that just didn't have anything in their Routing Table to give us  and hence, we have no peers  to talk to..we are starved.

Reply:
Code robustness thing (as discussed online): Having the starvation condition allows you to be able to vary the lookup condition without worrying that the lookup condition has to imply starvation.